extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java

```diff
@@ -54,54 +59,87 @@ public RecordReader<NullWritable, Object> createRecordReader(InputSplit split, T
-                Path file = fileSplit.getPath();
-                FileSystem fs = file.getFileSystem(conf);
-                FSDataInputStream in = fs.open(file);
-
+                formatClazz = className == null ? null : ReflectionUtils.loadClass(className);
+                String[] header = readHeader(fileSplit, context);
                 if (formatClazz == String[].class) {
-                    ObjectReader reader = new CsvMapper().enable(Feature.WRAP_AS_ARRAY)
-                                                         .readerFor(String[].class)
-                                                         .with(CsvSchema.emptySchema().withSkipFirstDataRow(false));
+                    objectReader = new CsvMapper()
+                            .enable(Feature.WRAP_AS_ARRAY)
+                            .readerFor(String[].class)
+                            .with(CsvSchema.emptySchema().withSkipFirstDataRow(false));
 
-                    iterator = reader.readValues((InputStream) in);
-                    if (!iterator.hasNext()) {
-                        throw new JetException("Header row missing in " + split);
-                    }
-                    String[] header = (String[]) iterator.next();
                     List<String> fieldNames = new ArrayList<>();
                     String field;
                     for (int i = 0; (field = configuration.get(CSV_INPUT_FORMAT_FIELD_LIST_PREFIX + i)) != null; i++) {
                         fieldNames.add(field);
                     }
+
                     projection = (Function) createFieldProjection(header, fieldNames);
                 } else {
-                    iterator = new CsvMapper().readerFor(formatClazz)
-                                              .withoutFeatures(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
-                                              .with(CsvSchema.emptySchema().withHeader())
-                                              .readValues((InputStream) in);
+                    Builder builder = CsvSchema.builder();
+                    for (String column : header) {
+                        builder.addColumn(column);
+                    }
+                    objectReader = new CsvMapper().readerFor(formatClazz)
+                                                  .withoutFeatures(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+                                                  .with(builder.build());
                 }
             }
 
-            @Override
-            public boolean nextKeyValue() {
-                if (!iterator.hasNext()) {
-                    return false;
+            private String[] readHeader(FileSplit fileSplit, TaskAttemptContext context) throws IOException {
+                Configuration configuration = context.getConfiguration();
+                Path file = fileSplit.getPath();
+                FileSystem fs = file.getFileSystem(configuration);
+
+                try (InputStream in = fs.open(file)) {
```

Comment:
you can merge those try blocks:
```java
ObjectReader headerReader = new CsvMapper()
                                .enable(Feature.WRAP_AS_ARRAY)
                                .readerFor(String[].class)
                                .with(CsvSchema.emptySchema().withSkipFirstDataRow(false));
try (
    InputStream in = fs.open(file);
    InputStream wrapped = wrap(in, configuration, file);
    MappingIterator<Object> iterator = headerReader.readValues(wrapped
) {
```

Reply:
Done in https://github.com/hazelcast/hazelcast/commit/ec0def77318fcb57eca1afa98af365acb984db8d